cocotb/types/array.py

```diff
@@ -0,0 +1,314 @@
+    ):
+        if value is not None and range is None:
+            self._value = self._construct_value(value)
+            self._range = self._construct_range(Range(0, "to", len(self._value) - 1))
+        elif value is not None and range is not None:
+            self._value = self._construct_value(value)
+            self._range = self._construct_range(range)
+            if len(self._value) != len(self._range):
+                raise ValueError(
+                    "init value of length {!r} not fit in {!r}".format(
+                        len(self._value), self._range
+                    )
+                )
+        elif value is None and range is not None:
+            self._value = self._construct_value(None for _ in range)
+            self._range = self._construct_range(range)
+        else:
+            raise TypeError("must pass a value, range, or both")
+
+    @staticmethod
+    def _construct_value(value: Iterable[Any]) -> List[Any]:
+        """
+        Constructs the value portion of the array.
+
+        For overriding by subclasses.
+        Used by constructor when values are given
+        and by __setitem__ when setting a slice to an iterable.
+
+        Args:
+            value: Any iterable of values that are constructable into elements using :meth:`_construct_elements`
+
+        Returns:
+            A :class:`list` of elements.
+
+        Raises:
+            TypeError: When the type isn't supported.
+            ValueError: When the iterable's values prevent construction into a list of elements.
+        """
+        return list(value)
+
+    @staticmethod
+    def _construct_element(elem: Any) -> Any:
+        """
+        Constructs a single element of an array.
+
+        For overriding by subclasses.
+        Used by __setitem__ when setting a single element.
+
+        Args:
+            elem: Any object that the implementation can turn into an array element.
+
+        Returns:
+            An array element.
+
+        Raises:
+            TypeError: When the type isn't supported.
+            ValueError: When the value prevents construction into an element.
+        """
+        return elem
+
+    @staticmethod
+    def _construct_range(rng: Any) -> Range:
+        """
+        Constructs the range portion of the array.
+
+        For overriding by subclasses.
+        Used by the constructor when setting the range
+        and by __setitem__ when creating a slice.
+
+        Args:
+            rng: Any object that the implementation can turn into a :class:`Range` object.
+
+        Returns:
+            The :class:`Range` object describing the array's indexing scheme.
+
+        Raises:
+            TypeError: When type isn't supported.
+            ValueError: When the value prevents construction into a range.
+        """
+        if isinstance(rng, Range):
+            return rng
+        raise TypeError("range argument must be of type 'Range'")
```

Comment:
I'd be tempted to leave these extensionality hooks to a future PR - it's hard to review extension points without an example of something extending them.

Reply:
Well `_construct_range` itself isn't really useful for overloading, but I already wanted to split it off into another function because I call it in multiple places in the constructor. `_construct_element` and `_construct_value` are of immediate use for `LogicArray`. 

```python
@staticmethod
def _construct_element(elem: Any) -> Any:
    return Logic(elem)

@staticmethod
def _construct_value(value: Iterable[Any]) -> List[Any]:
    return [self._construct_element(v) for v in value]
```
For automatic conversion of all values into `Logic`.